optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/employee/EmployeeSubform.java

```diff
@@ -0,0 +1,151 @@
+import org.optaplanner.openshift.employeerostering.gwtui.client.tenant.TenantStore;
+import org.optaplanner.openshift.employeerostering.shared.employee.Employee;
+import org.optaplanner.openshift.employeerostering.shared.skill.Skill;
+
+@Dependent
+@Templated
+public class EmployeeSubform implements TakesValue<Employee> {
+
+    @Inject
+    Validator validator;
+
+    @Inject
+    @AutoBound
+    private DataBinder<Employee> employeeModelDataBinder;
+
+    @Inject
+    private TenantStore tenantStore;
+
+    @Inject
+    @Bound(property = "name")
+    @DataField
+    private TextBox employeeNameTextBox;
+
+    @Inject
+    @Bound(property = "skillProficiencySet", converter = SkillConvertor.class)
+    @DataField
+    private SingleValueTagsInputExtension skillProficiencyList;
+
+    @PostConstruct
+    protected void initWidget() {
+        skillProficiencyList.init(employeeModelDataBinder);
+        skillProficiencyList.setItemValue(Skill::getName);
+        skillProficiencyList.setItemText(Skill::getName);
+        skillProficiencyList.reconfigure();
+        employeeModelDataBinder.getModel().setTenantId(tenantStore.getCurrentTenantId());
+    }
+
+    public Set<ConstraintViolation<Employee>> validate() {
+        return validator.validate(employeeModelDataBinder.getWorkingModel());
+    }
+
+    public void reset() {
+        employeeNameTextBox.setValue("");
+    }
+
+    public Promise<Employee> getIfValid() {
+        return new Promise<Employee>((res, rej) -> {
+            Set<ConstraintViolation<Employee>> validationErrorSet = validate();
+            if (validationErrorSet.isEmpty()) {
+                res.onInvoke(employeeModelDataBinder.getModel());
+            } else {
+                rej.onInvoke(validationErrorSet);
+            }
+        });
+
+    }
+
+    public void onAnyTenantEvent(@Observes TenantStore.TenantChange tenant) {
+        employeeModelDataBinder.getModel().setTenantId(tenantStore.getCurrentTenantId());
+    }
+
+    public void setSkillList(List<Skill> skillList) {
+        skillProficiencyList.setSkillSet(skillList);
+        skillProficiencyList.removeAll();
+        skillProficiencyList.setDatasets((Dataset<Skill>) new CollectionDataset<Skill>(skillList) {
+
+            @Override
+            public String getValue(Skill skill) {
+                return (skill == null) ? "" : skill.getName();
+            }
+        });
+        skillProficiencyList.reconfigure();
+    }
+
+    @Inject
+    public void setEmployeeModel(Employee employeeModel) {
+        employeeModelDataBinder.setModel(employeeModel);
+    }
+
+    @Override
+    public void setValue(Employee value) {
+        setEmployeeModel(value);
+    }
+
+    @Override
+    public Employee getValue() {
+        return employeeModelDataBinder.getModel();
+    }
+
+    public static class SingleValueTagsInputExtension extends SingleValueTagsInput<Skill> implements HasText {
```

Comment:
Okay, I think I understand why you use HasText here: because the supertype is already `HasValueChangeHandler<String>`, right?

IMO, I think you could simplify this (and remove the need for the external converter) by using encapsulation instead of inheritance.

Here is a sketch of a replacement:

    public class SkillValueTagsInputWrapper implements HasValue<Set<Skill>>, IsWidget {
      @Inject
      private SingleValueTagsInput<Skill> wrapped;
    
      @Override 
      public Widget asWidget() {
        return wrapped;
      }
      
      /*
       * Implement HasValue methods here, which defers to the SingleValueTagsInput change
       * handlers + conversion logic.
       * The logic could be simplified because you could set a tenant id as a field of this type.
       */
    }

This should work because `@DataField` respects `IsWidget` and will composite the object returned by `asWidget` onto the screen, but `@Bound` does not care about `IsWidget` and will respect the `HasValue` implementation of the wrapper class.

Reply:
Okay, I spoke too soon and doing IsWidget won't work, but there are still other ways you could try to encapsulate this.

You could make a `@Templated` component that has the `SingleValueTagsInput` as it's only element. It's a bit of boiler plate (and a silly one element HTML file) but I still think it is cleaner.

Reply:
You could also extend Composite, which is a widget type meant exactly for encapsulating and adding behaviour to a single Widget.